interop-2022/main.js

```diff
@@ -0,0 +1,477 @@
+'use strict';
+
+// TODO: There's a lot of reused code from browser-specific-failures.js here,
+// that could be put into lib/
+
+const fetch = require('node-fetch');
+const flags = require('flags');
+const fs = require('fs');
+const Git = require('nodegit');
+const lib = require('../lib');
+const moment = require('moment');
+const path = require('path');
+
+flags.defineStringList('products', ['chrome', 'firefox', 'safari'],
+    'Products to include (comma-separated)');
+flags.defineString('from', '2018-07-01', 'Starting date (inclusive)');
+flags.defineString('to', moment().format('YYYY-MM-DD'),
+    'Ending date (exclusive)');
+flags.defineBoolean('experimental', false,
+    'Calculate metrics for experimental runs.');
+flags.parse();
+
+const ROOT_DIR = path.join(__dirname, '..');
+
+const CATEGORIES = [
+  'interop-2021-aspect-ratio',
+  'interop-2021-flexbox',
+  'interop-2021-grid',
+  'interop-2021-position-sticky',
+  'interop-2021-transforms',
+  'interop-2022-cascade',
+  'interop-2022-color',
+  'interop-2022-contain',
+  'interop-2022-dialog',
+  'interop-2022-forms',
+  'interop-2022-scrolling',
+  'interop-2022-subgrid',
+  'interop-2022-text',
+  // 'interop-2022-viewport',
+  'interop-2022-webcompat',
+];
+
+// See documentation of advanceDateToSkipBadDataIfNecessary. These ranges are
+// inclusive, exclusive.
+const STABLE_BAD_RANGES = [
+  // This was some form of Safari outage, undiagnosed but a clear erroneous
+  // spike in failure rates.
+  [moment('2019-02-06'), moment('2019-03-04')],
+  // This was a safaridriver outage, resolved by
+  // https://github.com/web-platform-tests/wpt/pull/18585
+  [moment('2019-06-27'), moment('2019-08-23')],
+  // This was a general outage due to the Taskcluster Checks migration.
+  [moment('2020-07-08'), moment('2020-07-16')],
+  // This was a Firefox outage which produced only partial test results.
+  [moment('2020-07-21'), moment('2020-08-15')],
+];
+const EXPERIMENTAL_BAD_RANGES = [
+  // This was a safaridriver outage, resolved by
+  // https://github.com/web-platform-tests/wpt/pull/18585
+  [moment('2019-06-27'), moment('2019-08-23')],
+  // This was a general outage due to the Taskcluster Checks migration.
+  [moment('2020-07-08'), moment('2020-07-16')],
+  // Something went wrong with the Firefox run on this date.
+  [moment('2021-03-08'), moment('2021-03-09')],
+];
```

Comment:
Can we extract this into a new module, rather than having it copied across various scripts? Notably, this is out of date re: #76.

Reply:
Good idea, done in https://github.com/Ecosystem-Infra/wpt-results-analysis/pull/78.

Reply:
I'll cherry-pick that into this PR and rebase it away when #78 merged.